src/toil/bus.py

```diff
@@ -599,5 +623,84 @@ def __init__(self) -> None:
+
+    We track the state and name of jobs here, by ID.
+    We would use a list of two items but MyPy can't understand a list
+    of items of multiple types, so we need to define a new class.
+    """
+    @dataclass
+    class JobStatus:
+        """
+        Records the status of a job.
+        """
+
+        job_store_id: str
+        name: str
+        exit_code: int
+        annotations: Dict[str, str]
+        toil_batch_id: int
+        external_batch_id: str
+        batch_system: str
+
+        def toJSON(self) -> str:
+            return json.dumps(self, default= lambda o: o.__dict__, indent=4)
+
+    job_statuses: Dict[str, JobStatus] = collections.defaultdict(lambda: JobStatus('', '', -1, {}, -1, '', ''))
+    batch_to_job_id = {}
+    with open(path, 'rb') as log_stream:
+        # Read all the full, properly-terminated messages about job updates
+        for event in MessageBus.scan_bus_messages(log_stream, [JobUpdatedMessage, JobIssuedMessage, JobCompletedMessage,
+                                                               JobFailedMessage, JobAnnotationMessage, BatchSystemMessage]):
+            # And for each of them
+            logger.info('Got message from workflow: %s', event)
+
+            if isinstance(event, JobUpdatedMessage):
+                # Apply the latest return code from the job with this ID.
+                job_statuses[event.job_id].exit_code = event.result_status
+            elif isinstance(event, JobIssuedMessage):
+                job_statuses[event.job_id].job_store_id = event.job_id
+                job_statuses[event.job_id].name = event.job_type
+                job_statuses[event.job_id].toil_batch_id = event.toil_batch_id
+                job_statuses[event.job_id].exit_code = -1
+                batch_to_job_id[event.toil_batch_id] = event.job_id
+            elif isinstance(event, JobCompletedMessage):
+                job_statuses[event.job_id].name = event.job_type
+                job_statuses[event.job_id].exit_code = event.exit_code
+            elif isinstance(event, JobFailedMessage):
+                job_statuses[event.job_id].name = event.job_type
+                if job_statuses[event.job_id].exit_code == 0:
+                    # Record the failure if we never got a failed exit code.
+                    job_statuses[event.job_id].exit_code = 1
+            elif isinstance(event, JobAnnotationMessage):
+                # Remember the last value of any annotation that is set
+                job_statuses[event.job_id].annotations[event.annotation_name] = event.annotation_value
+            elif isinstance(event, BatchSystemMessage):
+                if event.toil_batch_id in batch_to_job_id:
+                    job_statuses[batch_to_job_id[event.toil_batch_id]].external_batch_id = event.external_batch_id
+                    job_statuses[batch_to_job_id[event.toil_batch_id]].batch_system = event.batch_system
+
+    return job_statuses
+
+def gen_messBus_path() -> str:
```

Comment:
This name isn't `snake_case`. I also don't think it's a good idea to abbreviate the message bus as `messBus`; we don't call this enough for that to be worth it.